apps/client/src/app/modules/list/data/ListRowSerializationHelper.ts

```diff
@@ -0,0 +1,315 @@
+interface ItemNameHolder { id: number | string, itemId?: string, name?: string, itemIdName?: string };
+interface NPCNameHolder { id: number, npcId?: number, npcName?: string, zoneName?: string };
+interface ZoneInfo { id?: number; npcId?: number; npcName?: string; zoneName?: string; zoneId?: any; zoneid?: any; mapId?: any; };
+interface MapInfo { zoneId?: any; zoneid?: any; mapId?: any; };
+interface CurrencyInfo { currencies: any[]; items: any[]; };
+interface NpcTradeData { npcs: any; trades: CurrencyInfo[]; };
+interface MonsterInfo { id: any; itemId?: any; position?: any; npcId?: number; npcName?: string; zoneName?: string; zoneId?: any; zoneid?: any; mapId?: any; };
+interface MonsterInfoWithMobName extends MonsterInfo, NameHolder { }
+interface VendorWithName extends Vendor, NameHolder { }
+interface InstanceWithName extends Instance, NameHolder { }
+interface GatherWithName extends GatheredBy, NameHolder { }
+
+export class ListRowSerializationHelper {
+
+  public static getData<T = any>(row: ListRow, type: DataType, isObject = false): T {
+    return getItemSource<T>(row, type, isObject);
+  }
+
+  constructor(
+    private i18nTools: I18nToolsService,
+    private l12n: LocalizedDataService,
+    private gt: GarlandToolsService,
+  ) {
+  }
+  // guesses the name based off potential paths
+  public applyItemName(obj: ItemNameHolder): ItemNameHolder {
+    const id = obj.id ? obj.id : obj.itemId ? obj.itemId : undefined;
+    return {
+      ...obj,
+      name: id ? this.getItemName(id) : undefined,
+      itemIdName: obj.itemId ? this.getItemName(obj.itemId) : undefined
+    };
+  }
+
+  private getItemName(id: any): string {
+    const itemInfo = this.l12n.getItem(id);
+    return id ? this.getNameIfExists(itemInfo) : undefined;
+  }
+
+  private serializeCraftedBy(craftedBy: any[]): any {
+    return craftedBy ? craftedBy.map((obj: any) => {
+      const retval = {
+        ...obj,
+        name: this.getItemName(obj.id),
+        job: this.getJobAbbreviationFromId(obj.job)
+      };
+      return retval
+    }) : undefined;
+  }
+  private getJobAbbreviationFromId(job: number): string {
+    const jobAbbr = this.getNameIfExists(this.l12n.getJobAbbr(job));
+    return jobAbbr != null && jobAbbr !== "missing name" ? jobAbbr : "DOWM";
+  }
+
+  private getNameIfExists(name: I18nName): string {
+    const translatedName = this.i18nTools.getName(name);
+    return translatedName && translatedName !== 'no name' ? translatedName : undefined;
+  }
+
+  private serializeVoyages(voyages: any): string {
+    return voyages && voyages.length > 0 ? voyages.map((r: any) => this.getNameIfExists(r)) : undefined;
+  }
+
+  public applyNpcName(obj: Vendor): any {
+    return {
+      ...obj,
+      npcName: this.getNameIfExists(this.l12n.getNpc(obj.npcId)),
+      zoneName: this.getZoneName(obj),
+    };
+  }
+
+  private getZoneName(obj: MapInfo): string {
+    if (!obj)
+      return undefined;
```

Comment:
Return type says string but you're returning undefined here, this means that I should be able to do `getZoneName().toLowercase()` for instance, based on the typing it wouldn't crash, but it can because in reality, the return type doesn't match what can really be returned.

Reply:
ah! so I should explicitly put `string|undefined` the intent is to not generate the key in the final JSON